### ISSUE TITLE
Quest completed immediately

### ISSUE THREAD

**Issue by logicbutton** [HabitRPG/habitica#5518]
A little over an hour ago, I sent an invitation for the Moonstone Chain quest to my party. I refreshed the page a bit later and saw this:
![image](https://cloud.githubusercontent.com/assets/13093573/8400643/a80a7f14-1de7-11e5-91c5-13e797f68fe3.png)

I received all the quest rewards, including the Part 2 scroll, and it's appearing in the list of completed quests of everyone in the party, including people who have been in the Inn for months, even though none of us has ever done this quest before. Also, I still have the original Moonstone Chain quest scroll in addition to the Part 2 scroll.

One thing that may or may not be relevant: I originally leveled up to level 60 by casting Burst of Flames, but got the skill level up scroll drop error. So I refreshed, unchecked some dailies, cast Burst of Flames again, rechecked the dailies, and levelled up normally and got the scroll.

**Comment by Alys** [comment]
Hi @logicbutton. To investigate this, we'll need to be able to identify user accounts and your party. Please go to Help -> Report a Bug and see the instructions there. Step 6 describes how you can find your own User ID and the Group ID. Please paste them into this issue.  Please also include the User ID of RachelG8489 which you can find in her profile by clicking on her name in the list of party members, or by clicking on her name in any chat message she has made.

**Comment by logicbutton** [comment]
Whoops, knew I was forgetting something!

My ID: d43f0d55-51cc-478f-ab9b-1ab76d29d2d5
RachelG8489's ID: e7282214-6e97-42e7-bd9e-e4569d3cb345
Group ID: e1b31c34-8270-42c8-af8f-9c6d212cabee

**Comment by Alys** [comment]
Duplicated with test users in a test party. The quest is completed as soon as one of the invited users logs in (presumably triggered by receiving the invitation).

![screen shot 2015-06-29 at 5 14 27 pm](https://cloud.githubusercontent.com/assets/1495809/8402841/5d5c5898-1e82-11e5-86ec-414533abf430.png)

**Comment by Alys** [comment]
Does not occur a second time with the same test party. I'm guessing there's a configuration mismatch between the old quest code and new code that was released today-ish. I'm investigating further.

**Comment by Alys** [comment]
@logicbutton Is it possible that RachelG8489 had not been active in HabitRPG for a while? Did she last log in BEFORE a previous quest finished, and then she logged in today after you sent the invitations for the new quest?

**Comment by Alys** [comment]
@logicbutton You can ignore any messages we type here unless they are directly addressed to you with your name. I need to record things that I find in the code and database in case they're useful for other devs who might be looking at this. Ignore the rest of this message. :)

[EDIT: this was irrelevant] The new quest code contains tests that look for `user.party.quest.completed` being null, and that delete `user.party.quest.completed`. However some users in the database have `user.party.quest.completed` set to `""` (empty string). I'm wondering if that's triggering the quest completion behaviour. Investigating further....

**Comment by Alys** [comment]
EDIT to previous message: `quest.completed` should read `user.party.quest.completed`

**Comment by Alys** [comment]
Code that ends a collection quest and prints the messages to party chat:
https://github.com/HabitRPG/habitrpg/blob/b72b14fe6c0496edf7e2920be3d473b558bce2e6/website/src/models/group.js#L201-233
but that function hasn't changed since February.

**Comment by Alys** [comment]
With my test party, I can make it happen a second time if I set `lastCron` for one of the invitees to be yesterday and then send the invitation before that user logs in. I.e., the problem happens if the user sees a new quest invitation during/after undergoing cron. So far I've just done this with a collection quest.

**Comment by Alys** [comment]
When I tried to do the same thing using a boss quest, this red error notification occurred:

**Please reload, "Error Cast to number failed for value "NaN" at path "quest.progress.hp""** If the error persists, please report it at Help > Report a Bug. If you're familiar with your browser's console, please include any error messages.

That happened twice with a boss quest.

When I tried again with a collection quest, the quest completed itself immediately again.

**I.e., this is the sequence of events:**
1. **A quest owner sends a quest invitation.**
2. **A user in the same party starts using HabitRPG for the first time today, triggering cron.**
3. 
   - **If the quest is a boss quest, that user sees the "Cast to number failed" error.**
   - **If the quest is a collection quest, the quest immediately starts and finishes, with that user finding " " items (see screenshot at the top of this issue). The user never sees the quest invitation modal. The user does see the quest completed modal.**

This is quite buggy, and there's an unrelated (not serious) problem with closing notifications too (https://github.com/HabitRPG/habitrpg/issues/5519). If I can't solve this quest problem in the next few minutes, I'm reverting the latest deploy.

**Comment by Alys** [comment]
Note that for the user who triggers the quest completion, the quest invitation modal never appears. All they see is the quest completed modal.

**Comment by Alys** [comment]
This is the user's `party.quest` object after the invitation has been sent and before the user logs in to trigger cron and the quest completion:

```
    "party" : {
        "quest" : {
            "key" : "moonstone1",
            "progress" : {
                "up" : 0,
                "down" : 0,
                "collect" : {

                }
            },
            "completed" : "",
            "RSVPNeeded" : true
        }
    }
```

**Comment by Alys** [comment]
Yeah, I'm just flailing about here. I've reverted production to the previous version. I'll keep looking at the code though because I DO NOT WANT IT TO BEAT ME.

**Comment by Alys** [comment]
@logicbutton Thank you VERY much for telling us about this! We don't yet know exactly why it happened, but we will sort it out.

I'm sorry your party was affected by it. If you'd like me to revert any of the incorrect changes that happened to your party, just say so. For example, I can remove the quest achievement from all party members and/or remove the quest rewards. Of course if you'd rather keep them, that's fine too. :) Just let me know what, if anything, you'd like me to do.

**Comment by Alys** [comment]
At the point where this bug happens (invitation sent, quest NOT started), [this isOnQuest line](https://github.com/HabitRPG/habitrpg/blob/b72b14fe6c0496edf7e2920be3d473b558bce2e6/website/src/models/group.js#L203) should be preventing the rest of the code in that function from executing (the same line occurs in the [boss quest code](https://github.com/HabitRPG/habitrpg/blob/b72b14fe6c0496edf7e2920be3d473b558bce2e6/website/src/models/group.js#L304-L343) which is where the `Cast to number failed for value "NaN" at path "quest.progress.hp"` error occurs). 

So, the [isOnQuest function](https://github.com/HabitRPG/habitrpg/blob/b72b14fe6c0496edf7e2920be3d473b558bce2e6/website/src/models/group.js#L197-L199) is failing - it's reporting that the quest has started when it hasn't. Is `user.party.quest.key` being set prematurely?

**Comment by Alys** [comment]
Oh wait, [yes it is](https://github.com/HabitRPG/habitrpg/issues/5518#issuecomment-116504192). I should read my own notes. :) 

That's where the bug lies, I think.

**Comment by Alys** [comment]
And [this is where user.party.quest.key is being set](https://github.com/HabitRPG/habitrpg/blob/b72b14fe6c0496edf7e2920be3d473b558bce2e6/website/src/controllers/groups.js#L925) when the invitations are sent, so we just need to work out why that's being done, is it necessary, and if so, what can we put in the [isOnQuest function](https://github.com/HabitRPG/habitrpg/blob/b72b14fe6c0496edf7e2920be3d473b558bce2e6/website/src/models/group.js#L197-L199) to correctly check for whether the quest has started.

**Comment by Alys** [comment]
I'm not sure if user.party.quest.key needs to be set at that point, but regardless, it's not a good way to check for the quest being active. Checking for `group.quest.active` seems more reliable. :)

The PR fixes this problem on my local install. If it's merged when I'm not around, someone should repeat [this test](https://github.com/HabitRPG/habitrpg/issues/5518#issuecomment-116497644) before merging to ensure they can replicate the bug, and then again after. I'll be online again in about 12 hours.

**Comment by crookedneighbor** [comment]
Ahh, so the reason we didn't catch this in our testing is because the PR was specifically so you didn't have to refresh the page to get the quest invitation... so we must not have bother testing cron. Whoops :disappointed:

**Comment by hairlessbear** [comment]
@Alys `user.party.quest.key` is being set on quest invite so the rewards section of the invite modal will display properly. Previously, it relied on `party.quest.key`, but since the modal now displays on user sync instead of party sync, the modal can pop-up _before_ party sync, causing `party.quest.key` to be unreliable. So I set `user.party.quest.key` on quest invite, and changed the rewards modal to use `user.party.quest.key` instead.

I should've looked into whether that caused any conflicts with the rest of the code...sorry :confused:

**Comment by crookedneighbor** [comment]
Don't sweat it @hairlessbear. These things happen. I've done much much worse :smile:

**Comment by logicbutton** [comment]
@Alys Thank YOU (and everyone in this thread) for your prompt response! Wish my support tickets at work would get answered so quickly. :) I've asked the party what changes they'd like made; will report back.

**Comment by hairlessbear** [comment]
@crookedneighbor I'll keep that in mind; thanks :sweat_smile:

**Comment by logicbutton** [comment]
@Alys Could you please remove the quest achievement from everyone's profiles, but leave the rewards?

Thanks again for your help with this!

**Comment by Alys** [comment]
@logicbutton I've removed the quest achievement from all party members. Please tell me if anyone reports problems with that!

**Comment by jylyn** [comment]
I see this is closed but I've just had the same thing happen to me.  My browser crashed when I hit the sync button this morning and I came back to see:

TangyDragonBBQ found NaN Testimonies. 
All items found! Party has received their rewards.

My cron was the first in the party to run after the quest started.  I'm using the the beta server and I probably haven't refreshed the tab in a few days.

My ID: 987c6d4f-72f2-49a7-888a-e853b11eaff7
Group ID: ae2724bf-6f6f-4afd-b518-444f36604acf

**Comment by hairlessbear** [comment]
@jylyn Two things from your message jump out immediately.
1. "I'm using the beta server"
2. "I probably haven't refreshed the tab in a few days"

Either or both of those could have caused this. First of all, you shouldn't be using the beta site, at all. It's there for the devs to test out new features, and as such, is often unstable and contains bugs. Additionally, it doesn't necessarily have the same bug fixes that production does, meaning that although a fix for this was pushed to the production site, it may not have been pushed to the beta site.

Regarding refreshing: if you don't refresh the page every once in a while, you don't get new code, including bug fixes. The fix that @Alys wrote was merged in less than 36 hours ago, so even if it has been put on the beta site, you don't have it on your machine. 

Please stop using the beta site, and try to refresh the page at least once a day. That should help avoid problems like this in the future. If you have any additional questions, feel free to ask and I'll do my best to answer them :)

What would you like done about the accidentally-completed quest? I'm not an admin, so I can't make any changes, but one of them will be by sooner or later and can help you out.

**Comment by jylyn** [comment]
Thanks @hairlessbear - I figured it was a combination of those things that was to blame!  I was using the beta site because I kept getting killed from #4313 but perhaps it's time to switch back.

Just thought I would also point out that this happened AFTER I accepted the quest invitation - it seems from reading the messages above that it was previously happening to users who hadn't accepted.  Not sure if that makes a difference or not...

**Comment by Alys** [comment]
@jylyn Exactly as @hairlessbear says, please don't use the beta site. We can't consider this to be a bug that needs fixing since beta is currently being used for testing a specific feature that is not related to quests. If you use the main site, this will not happen.

If the quest participants want the quest achievement to be removed, I am willing to do that. Please provide a list of UUIDs for all affected players.

**Comment by hairlessbear** [comment]
There is still no fix for #4313, especially not on the beta site. But one way to mitigate the effects of #4313 is to refresh your page every once in a while. You may have been more affected by the bug due to leaving the tab open for multiple days at a time.
